### PR TITLE
Fix timestamp <-> slot calculations and add tests

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

Fixes the timestamp and slot conversion calculations given TIP updates.

### CODE REVIEW

File: sdk/src/types/block/slot/index.rs [hunk 1]
```diff
@@ -43,13 +43,22 @@ impl SlotIndex {
...
+        timestamp
+            .checked_sub(genesis_unix_timestamp as u64)
+            .map(|diff| (diff / slot_duration_in_seconds as u64) + 1)
+            .unwrap_or_default()
+            .into()
     }
 
-    /// Converts the slot index into the corresponding unix timestamp.
+    /// Converts the slot index into the unix timestamp representing the beginning of the slot.
+    /// Slot `0` will return the unix epoch.
```
Comment (thibault-martinez) [inline sdk/src/types/block/slot/index.rs:56]: Don't really have a choice but feels a tiny bit weird

**Review by thibault-martinez — APPROVED** [review]

File: sdk/src/types/block/slot/index.rs [hunk 2]
```diff
@@ -43,13 +43,22 @@ impl SlotIndex {
     }
 
     /// Gets the slot index of a unix timestamp.
+    /// Slots are counted starting from `1` with `0` being reserved for times before the genesis.
```
Comment (Thoralf-M) [inline sdk/src/types/block/slot/index.rs:46]: This sentence got updated in https://github.com/iotaledger/tips-draft/pull/20/commits/e95818e621e3edbcc6693b110024781bf5631a3f

Comment (DaughterOfMars) [reply sdk/src/types/block/slot/index.rs:46]: I'm not sure what's wrong?

Comment (Thoralf-M) [reply sdk/src/types/block/slot/index.rs:46]: Nothing wrong, just wasn't sure if you maybe want to add the other part, but also fine without it

**Review by Thoralf-M — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

File: sdk/src/types/block/slot/index.rs [hunk 3]
```diff
@@ -61,3 +70,51 @@ impl From<SlotIndex> for u64 {
 
 #[cfg(feature = "serde")]
 string_serde_impl!(SlotIndex);
+
+#[cfg(test)]
+mod test {
+    use crate::types::block::protocol::ProtocolParameters;
+
+    #[test]
+    fn to_from_timestamp() {
```
Comment (abdulmth) [inline sdk/src/types/block/slot/index.rs:79]: this doesn't test `from`

**Review by abdulmth — APPROVED** [review]

File: sdk/src/types/block/slot/index.rs (hunk 3 above)
Comment (thibault-martinez) [reply sdk/src/types/block/slot/index.rs:79]: `protocol_params.slot_index(timestamp);` calls from

**Pull request merged**